Row-first option for columnize

naught101 opened the pull request:
This adds a row-first sort to the columnize function, to address part of #8741 .

Original columnise: 

``` python
In [9]: print(columnize(items))
a  c  eeeeeeeeeeeeeeeeeeeeeee  eeeeeeeeeeeeeeeeeeeeeee
b  d  eeeeeeeeeeeeeeeeeeeeeee  eeeeeeeeeeeeeeeeeeeeeee
```

New addition:

``` python
In [10]: print(columnize(items, row_first=True))
a                        b                        c                      
d                        eeeeeeeeeeeeeeeeeeeeeee  eeeeeeeeeeeeeeeeeeeeeee
eeeeeeeeeeeeeeeeeeeeeee  eeeeeeeeeeeeeeeeeeeeeee
```

naught101 committed:
Row-first option for columnize

Carreau commented on IPython/utils/text.py:
```diff
@@ -714,6 +725,10 @@ def columnize(items, separator='  ', displaywidth=80):
         The strings to process.
 
+    row_first : (default False)
+        Whether to to compute columns for a row-first matrix instead ofr
```
instead `ofr` :-)

naught101 replied:
also, 'to to'.. gah..

Carreau commented:
Looks good to me. 

Want to add test for the option in `IPython/utils/tests/test_text.py` ?

naught101 committed:
fix help text

naught101 committed:
add some extra tests for row-first columnize

naught101 committed:
Fix loop logic

naught101 committed:
rename num_rows etc to max_rows for clarity

It can be the case when calculating row-first that the number of columns ends up allowing all cells to be accomodated in fewer rows. For example, if you have 6 items, and set nrows=4, then ncols=2, but 6 items can be accomodated in 3 rows.

naught101 commented:
Guess that was a good idea :)

I think it's ready to go.

Carreau commented:
> Guess that was a good idea :)

:-)

naught101 commented:
Hrm.. thoughts please:

currently, the row-first option calculates the optimal layout by starting with the max number of rows, then calculating the minimum number of columns, which may end up being much smaller than the display width. For example, this is the current result:

```{python}
In [3]: [s*15 for s in 'abcdef']
Out[3]: 
['aaaaaaaaaaaaaaa', 'bbbbbbbbbbbbbbb', 'ccccccccccccccc'
 'ddddddddddddddd', 'eeeeeeeeeeeeeee', 'fffffffffffffff']
```

Another option would be to maximize the columns, while remaining under the display limit. which would look like this:

```{python}
In [3]: [s*15 for s in 'abcdef']
Out[3]: 
['aaaaaaaaaaaaaaa', 'bbbbbbbbbbbbbbb', 'ccccccccccccccc', 'ddddddddddddddd'
 'eeeeeeeeeeeeeee', 'fffffffffffffff']
```

which might look nicer in some cases. However, it would never be more efficient in terms of space used, and it would require somewhat more complex code. 

Do you care which option is used?

Carreau commented:
I think the point of solution 1 was to make completion always closer to the cursor if you wrap around, and as the qtconsole is now it's own package, columnize should  not even used anymore (i grepped quickly but you can re-check)... so feel free to refactor any way you want.

Carreau commented:
We can also merge that and you can update the algo in another PR.

naught101 committed:
Add spread option, to allow columns to fill space.

naught101 commented:
Ok, I just had a though, and added a 'spread' option, which makes columnize spread the columns out to the maximum allowable width (optimal spacing was already calculated, I just left justified the separator to match the padding). 4 Tests included.

As far as I am concerned, this is ready to go - good enough for the pretty print issue anyway.

Carreau commented:
Ok, let's move forward then
